libraries/AP_BattMonitor/AP_BattMonitor_ACS37800.cpp

```diff
@@ -0,0 +1,187 @@
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "AP_BattMonitor_ACS37800.h"
+#include <GCS_MAVLink/GCS.h>
+#include <AP_HAL/utility/sparse-endian.h>
+#include <stdio.h>
+extern const AP_HAL::HAL& hal;
+
+typedef struct
+{
+  union
+  {
+    uint32_t all;
+    struct
+    {
+      uint32_t vcodes : 16;
+      uint32_t icodes : 16;
+    } bits;
+  } data;
+} acs378002a_t;
+
+#define RESIST_MULT 202.5
+
+#define REG_ACCESS_CODE 0x2F // access_code
+#define CUSTOMER_CODE 0x4F70656E
+#define REG_C_SHADOW 0x1C
+#define REG_F_SHADOW 0x1F
+#define REG_F_EEPROM 0x0F
+#define CONFIG_C_MASK 0x0000000F
+#define CONFIG_F 0x01FFC000 // bypass_N_en=1, n=1023
+
+#define REG_VRMS_IRMS 0x20 
+#define REG_VRMS_IRMS_ONESEC 0x26
+#define REG_VCODES_ICODES 0x2A
+#define REG_PINSTANT 0x2C
+
+#ifndef HAL_BATTMON_ACS37800_BUS
+#define HAL_BATTMON_ACS37800_BUS 0
+#endif
+#ifndef HAL_BATTMON_ACS37800_ADDR
+#define HAL_BATTMON_ACS37800_ADDR 96
+#endif
+
+const AP_Param::GroupInfo AP_BattMonitor_ACS37800::var_info[] = {
+
+    // @Param: I2C_BUS
+    // @DisplayName: Battery monitor I2C bus number
+    // @Description: Battery monitor I2C bus number
+    // @Range: 0 4
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("I2C_BUS", 25, AP_BattMonitor_ACS37800, i2c_bus, HAL_BATTMON_ACS37800_BUS),
+
+    // @Param: I2C_ADDR
+    // @DisplayName: Battery monitor I2C address
+    // @Description: Battery monitor I2C address
+    // @Range: 0 127
+    // @User: Advanced
+    // @RebootRequired: True
+    AP_GROUPINFO("I2C_ADDR", 26, AP_BattMonitor_ACS37800, i2c_address, HAL_BATTMON_ACS37800_ADDR),
+
+    // @Param: V_FINE_M
+    // @DisplayName: Voltage fine adjustment
+    // @Description: Fine-adjust the voltage reading. Calculate dividing the indicated measured voltage by the known applied voltage.
+    // @User: Advanced
+    AP_GROUPINFO("V_FINE_M", 27, AP_BattMonitor_ACS37800, volt_fine_mult, 1.0),
+
+    // @Param: C_FINE_M
+    // @DisplayName: Current fine adjustment
+    // @Description: Fine-adjust the current reading. Calculate dividing the indicated measured current by the known current draw.
+    // @User: Advanced
+    AP_GROUPINFO("C_FINE_M", 28, AP_BattMonitor_ACS37800, curr_fine_mult, 1.0),
+
+    AP_GROUPEND
+};
+
+AP_BattMonitor_ACS37800::AP_BattMonitor_ACS37800(AP_BattMonitor &mon, 
+                            AP_BattMonitor::BattMonitor_State &mon_state, 
+                            AP_BattMonitor_Params &params):
+        AP_BattMonitor_Backend(mon, mon_state, params)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    _state.var_info = var_info;
+}
+
+void AP_BattMonitor_ACS37800::init(void)
+{
+    _dev = hal.i2c_mgr->get_device(i2c_bus, i2c_address, 100000, false, 20);
+    if (!_dev) {
```

Comment:
```suggestion
    if (_dev == nullptr) {
```

... it's a style thing